src/database/redis.rs

```diff
@@ -0,0 +1,110 @@
+use super::models::DatabaseError;
+use deadpool_redis::{Config, Runtime};
+use redis::{cmd, FromRedisValue, ToRedisArgs};
+use std::fmt::Display;
+
+const DEFAULT_EXPIRY: i64 = 1800; // 30 minutes
+
+#[derive(Clone)]
+pub struct RedisPool {
+    pool: deadpool_redis::Pool,
+    meta_namespace: String,
+}
+
+impl RedisPool {
+    // initiate a new redis pool
+    // testing pool uses a hashmap to mimic redis behaviour for very small data sizes (ie: tests)
+    // PANICS: production pool will panic if redis url is not set
+    pub fn new(meta_namespace: Option<String>) -> Self {
+        let redis_pool = Config::from_url(dotenvy::var("REDIS_URL").expect("Redis URL not set"))
+            .builder()
+            .expect("Error building Redis pool")
+            .max_size(
+                dotenvy::var("DATABASE_MAX_CONNECTIONS")
+                    .ok()
+                    .and_then(|x| x.parse().ok())
+                    .unwrap_or(10000),
+            )
+            .runtime(Runtime::Tokio1)
+            .build()
+            .expect("Redis connection failed");
+
+        RedisPool {
+            pool: redis_pool,
+            meta_namespace: meta_namespace.unwrap_or("".to_string()),
+        }
+    }
+
+    pub async fn set<T1, T2>(
+        &self,
+        namespace: &str,
+        id: T1,
+        data: T2,
+        expiry: Option<i64>,
+    ) -> Result<(), DatabaseError>
+    where
+        T1: Display,
+        T2: ToRedisArgs,
+    {
+        let mut redis_connection = self.pool.get().await?;
```

Comment:
This was something I thought a little about, but I ended up not really worrying about it. Most areas that call redis currently only do 1-2 commands on it anyway per connection, so I don't really think it would affect that much either positively or negatively. 

I don't think it would be entirely too difficult to set it up that way, aside from some mild structure complication (ie: another struct for RedisConnection) if we think it would be an issue. @Geometrically thoughts?

Reply:
Yeah, I'm inclined to agree that it's fine, especially since when we have two we're usually doing a slow operation (database call) in between.